include/heapredirect.hpp

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#ifndef HEAPREDIRECT_H
+#define HEAPREDIRECT_H 
+
+#include "staticmutex.hpp"
+#include "staticbufferheap.hpp"
+
```

Comment:
It pains me for this whole thing to be a giant macro... if more of the code could be factored out and only a tiny bit actually be in a macro, that would be better.

Reply:
Sure, can I move some of that to a `.cpp` or would you prefer a `class` template?  I had it in a class template before, but that had some downsides (most easily remedied with C++17's `inline static` once again).

We do have to emit the `xxmalloc`, `xxfree`, etc. Because the various `Heap`s lack a common interface (base and `virtual` methods), we can't use a factory method for the heap and move them to a `.cpp`... it's easier to just leave them in the macro, even if all they do is invoke class template methods.

Let me know and I'll be happy to adjust.

Reply:
Yeah, I think a macro to emit the `xxmalloc` things is fine; class template also best IMHO (to the extent that this can be header-only (mostly), better).